### ISSUE TITLE
Manage instances of google_compute_instance_group in non-authoritative way

### ISSUE THREAD

**Issue by kponichtera** [hashicorp/terraform-provider-google#8605]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Let's take the following example, where we have a virtual machine instance and the instance group:

```tf
// Singleton disk with important data - cannot be shared between many machines
resource "google_compute_disk" "disk" {
  name = "disk"
  // rest skipped for brevity
}

resource "google_compute_instance" "instance" {
  name = "my-instance"
 
  attached_disk {
    source = google_compute_disk.disk.name
    device_name = "disk"
  }

  metadata_startup_script = templatefile("${path.module}/init.sh", {
    // variables skipped for brevity
  })

  // rest skipped for brevity
}

resource "google_compute_instance_group" "instance_group" {
  name = "instance-group"

  instances = [
    google_compute_instance.instance.self_link
  ]

  named_port {
    name = "http"
    port = "80"
  }

}

```

Right now managing the instances of `google_compute_instance_group` is done in the authoritative fashion, where the `google_compute_instance` IDs or self links are provided to the instance group `instances` field. However, when the virtual machine is recreated (which happens very often if somebody relies on the startup script), the ID and self link do not change and have a form of `projects/<project_id>/zones/<zone>/instances/<instance_name>` and `https://www.googleapis.com/compute/v1/projects/<project_id>/zones/<zone>/instances/<instance_name>`. Terraform realizes that the passed instance name _might_ change and shows it in the plan:

```tf
  # module.google_compute_instance_group.instance_group will be updated in-place
  ~ resource "google_compute_instance_group" "instance_group" {
        id        = "projects/project/zones/europe-west3-a/instanceGroups/instance"
      ~ instances = [
          - "https://www.googleapis.com/compute/v1/projects/project/zones/europe-west3-a/instanceGroups/instance",
        ] -> (known after apply)
        name      = "instance-group"
        # (5 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

However, the only thing the plan execution does is recreation of the virtual machine instance - the instance group itself is not modified in any way, simply because the instance self link it was supposed to change turned out to be the same. What's more, removing the virtual machine removed it from the instance group as well, and after the plan execution we end up with an empty instance group in Google Cloud. Interestingly, in the Terraform state the instance group resource still has the instance seemingly attached. In order to reattach the newly created virtual machine to the instance group it's necessary to perform reconciliation between Terraform state and the cloud, so that Terraform knows that the instance group is actually empty, which effectively requires one to execute `terraform apply` twice.

Recreation of the instance group with the use of `random_id` resource is not a viable option, since it triggers the `resourceInUseByAnotherResource` error, while specifying the `lifecycle {create_before_destroy = true}` option leads to creation of a second instance of the virtual machine, which cannot be done - not only the name of the virtual machine is already in use, but the machine itself is a singleton which uses specific instance of `google_compute_disk`. What's more, I cannot use the `random_id` trick here, since in order to regenerate it, I'd have to somehow inform the resource that the virtual machine instance has been recreated. The `google_compute_instance` resource has a nice little field called `instance_id` which is a long integer, assigned by the server, which changes every time the instance is created, but it cannot be used neither in the instance group list, nor for triggering `random_id` regeneration - that would introduce a cyclic dependency.

I know it's not really how the instance groups should be used, but it's the only way I could attach single virtual machine to the HTTPS load balancer - I don't want to expose it to the internet directly and I want to leverage Google's SSL certificate management and provisioning.

**TL;DR** Since it's possible to create empty instance group, it would be great to manage its virtual machine instances [in the same way the IAM policies are managed](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam) - through an extra resource, which binds the instance group with the actual instances, much like many-to-many relations in the database.

### New or Affected Resource(s)

* [google_compute_instance](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance)
* [google_compute_instance_group](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance_group)

### Potential Terraform Configuration

The proposed resource could be very simple, merely connecting the instance and the instance group:

```tf
resource "google_compute_instance_group_member" "instance_group_member" {
  instance_group = google_compute_instance_group.instance_group.self_link
  instance = google_compute_instance.instance.self_link
}
```

The creation of the resource could trigger adding it to the already existing instance group, while the removal would remove it from the group. Thanks to that, the order of operations would be as follows:

1. remove binding between instance group and virtual machine
2. remove virtual machine
3. create virtual machine
4. create binding between instance group and virtual machine

### References

* b/293598488
